### PR TITLE
Do not rely on file extensions after path canonicalization.

### PR DESCRIPTION
**Pull request by vadimcn**
Rustc does not recognize libraries which are symlinked to files having extension other than .rlib. The problem is that find_library_crate calls fs::canonicalize on found library paths, but then the resulting path is passed to get_metadata_section, which assumes it will end in ".rlib" if it's an rlib (from https://internals.rust-lang.org/t/is-library-path-canonicalization-worth-it/3206).

cc #29433

### CODE REVIEW

**Comment by vadimcn** [comment]
Also, I'd like to nominate this fix for backporting into 1.8.

File: src/librustc_metadata/loader.rs [hunk 1]
```diff
@@ -283,6 +284,21 @@ pub struct CratePaths {
...
+enum CrateFlavor {
+    Rlib,
+    Dylib
+}
+
+impl fmt::Display for CrateFlavor {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_str(match *self {
+            CrateFlavor::Rlib => "rlib",
+            CrateFlavor::Dylib => "dylib"
```
Comment (GuillaumeGomez) [inline src/librustc_metadata/loader.rs]: Could you add a comma at the end of this line please?

**Comment by alexcrichton** [comment]
This seems pretty similar to https://github.com/rust-lang/rust/pull/32317 in scope, but perhaps somewhat different in terms of problem solved? Could you also be sure to add a test for this as well?

**Comment by vadimcn** [comment]
No, I think it's the same issue.   Wasn't aware #32317 existed.

**Comment by alexcrichton** [comment]
I think that #32317 is arguably more general, but I somewhat prefer this tactic as it solves the problem at hand without changing the semantics that much

**Comment by alexcrichton** [comment]
cc @taralx

**Comment by vadimcn** [comment]
Added a test in case you want to go with this one.

**Comment by taralx** [comment]
This fixes the problem that I have specifically, although #32317 is probably also worth doing long-term (once I work out what's broken about Windows with it).

**Comment by alexcrichton** [comment]
@bors: r+ cc3b6f24b7604880279918b3c1a58ef62fb06b36

Ok, thanks @vadimcn! @taralx yeah I think we'll want to take both strategies, so I'll merge this for now while you're working on #32317, and we can merge once the Windows issue is tracked down.

**Comment by eddyb** [comment]
Nominating for backport on behalf of @taralx.

**Comment by nikomatsakis** [comment]
Discussed in @rust-lang/compiler meeting. Is this fixing a **regression** or just some longstanding bug?

**Comment by taralx** [comment]
This is technically a regression due to https://github.com/rust-lang/rust/pull/12474. Before that, symlinks named *.rlib worked, regardless of the name of the thing they pointed to.

**Comment by alexcrichton** [comment]
@nikomatsakis I believe, (ias @taralx pointed out) that this is just fixing a longstanding bug

**Comment by nikomatsakis** [comment]
@alexcrichton 

> I believe that this is just fixing a longstanding bug

In that case, I believe (based on consensus from previous @rust-lang/compiler meeting) we would prefer not to backport.

**Comment by vadimcn** [comment]
@rust-lang/compiler: Can you please reconsider?  This prevents Rust builds from working on our build cluster.  I believe this is a pretty non-invasive change, and the beta cycle has just started....

**Pull request merged**